_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,606 @@
+    x + y;
+    x.add(y); // ERROR: Member "add" not found or not visible after argument-dependent lookup in Int.
+
+    x - y;
+    x.sub(y); // OK -- "sub" was also attached in "using for"
+}
+```
+
+In the above example, the function `add()` is used *only* as the operator `+`, and not callable as `x.add(y)`
+(but it can, of course, be also called as `add(x, y)`).
+`sub()`, on the other hand, can both be used as the operator `-` and called on the type.
+
+The rules for `using for global` with operators are the same as for attached functions
+(with an additional restriction that operators *must* be global):
+it is only allowed at file level, the type must be defined in the same file and it makes
+the operator available everywhere.
+Note that only the directive itself must be colocated with the type.
+The operator definitions may be imported from a separate file.
+
+#### More Examples
+
+##### Unchecked Counter
+This example shows how operators can be used to bypass the checked arithmetic for types
+specifically defined for cases where those checks are superfluous:
+
+```solidity
+pragma solidity ^0.8.19;
+
+type UncheckedCounter is uint;
+
+using {add as +, lt as <} for UncheckedCounter global;
+
+UncheckedCounter constant ONE = UncheckedCounter.wrap(1);
+
+function add(UncheckedCounter x, UncheckedCounter y) pure returns (UncheckedCounter) {
+    unchecked {
+        return UncheckedCounter.wrap(
+            UncheckedCounter.unwrap(x) +
+            UncheckedCounter.unwrap(y)
+        );
+    }
+}
+
+function lt(UncheckedCounter x, UncheckedCounter y) pure returns (bool) {
+    return UncheckedCounter.unwrap(x) < UncheckedCounter.unwrap(y);
+}
+
+contract C {
+    uint internalCounter = 12;
+
+    function testCounter() public returns (uint) {
+        for (
+            UncheckedCounter i = UncheckedCounter.wrap(12);
+            i < UncheckedCounter.wrap(20);
+            i = i + ONE
+        )
+            ++internalCounter;
+
+        return internalCounter;
+    }
+}
+```
+
+##### Token Conversion
+This is a more complex example that shows multiple aspects of user-defined operators.
+The `TokenConversion` contract holds balances of two different tokens, defined as UDVTs,
+and uses a third type as a common denominator for calculations.
+
+`TokenConversion.sol`
+```solidity
+pragma solidity ^0.8.19;
+
+import {TokenA, TokenB, BaseToken, ZERO, ONE} from "./Tokens.sol";
+
+contract TokenConversion {
+    mapping(address account => TokenA balance) public balanceOfTokenA;
+    mapping(address account => TokenB balance) public balanceOfTokenB;
+
+    BaseToken averageDepositSize = ZERO;
+    BaseToken numberOfDeposits = ZERO;
+
+    function deposit(address receiver, TokenA amount) public {
+        balanceOfTokenA[receiver] = balanceOfTokenA[receiver] + amount;
+        BaseToken depositSize = amount.toBase();
+        averageDepositSize = (numberOfDeposits * averageDepositSize + depositSize) / (numberOfDeposits + ONE);
```

Comment:
The bad thing about this is that UDVTs should allow us to avoid using the same type for a plain number and a deposit size. I don't see who we can do better though. We can't have operators on mixed types.

Reply:
Yeah, I felt that need (operators on mixed types) here.

Reply:
One thing you could do would be to use attached functions in places where you would use such operators. We should actually show that off for `exp()`.

Reply:
I replaced the `+ ONE` operation with an attached function `increment`. 
Not sure how we would show that for `**`... or do you mean, we should come up with some function that makes use of `**` for UDVTs?